x-pack/plugin/mapper-aggregate-metric/src/main/java/org/elasticsearch/xpack/aggregatemetric/mapper/AggregateDoubleMetricFieldMapper.java

```diff
@@ -508,6 +511,117 @@ public ValueFetcher valueFetcher(SearchExecutionContext context, String format)
+
+                if (minValues == null || maxValues == null || sumValues == null || valueCountValues == null) {
+                    throw new UnsupportedOperationException("Must have all subfields to use aggregate double metric in ESQL");
+                }
+                return new BlockDocValuesReader() {
+
+                    private int docID = -1;
+
+                    @Override
+                    protected int docId() {
+                        return docID;
+                    }
+
+                    @Override
+                    public String toString() {
+                        return "BlockDocValuesReader.AggregatedDoubleMetrics";
+                    }
+
+                    @Override
+                    public Block read(BlockFactory factory, Docs docs) throws IOException {
+                        try (var builder = factory.aggregateDoubleMetricBuilder(docs.count())) {
+                            int lastDoc = -1;
+                            for (int i = 0; i < docs.count(); i++) {
+                                int doc = docs.get(i);
+                                if (doc < lastDoc) {
+                                    throw new IllegalStateException("docs within same block must be in order");
+                                }
+                                if (minValues.advanceExact(doc)) {
+                                    boolean found = maxValues.advanceExact(doc);
+                                    assert found;
+                                    found = sumValues.advanceExact(doc);
+                                    assert found;
+                                    found = valueCountValues.advanceExact(doc);
+                                    assert found;
+                                    double min = NumericUtils.sortableLongToDouble(minValues.longValue());
+                                    double max = NumericUtils.sortableLongToDouble(maxValues.longValue());
+                                    double sum = NumericUtils.sortableLongToDouble(sumValues.longValue());
+                                    int count = Math.toIntExact(valueCountValues.longValue());
+                                    builder.append(min, max, sum, count);
+                                } else {
+                                    builder.appendNull();
+                                }
+                                lastDoc = doc;
+                                this.docID = doc;
+                            }
+                            return builder.build();
+                        }
+                    }
+
+                    @Override
+                    public void read(int docId, StoredFields storedFields, Builder builder) throws IOException {
+                        this.docID = docId;
+                        var blockBuilder = (BlockLoader.AggregateDoubleMetricBuilder) builder;
```

Comment:
Can we create a `read(int docId, Builder builder)` then delegate these two reads to the new method?